packages/hoppscotch-js-sandbox/src/utils.ts

```diff
@@ -104,14 +111,20 @@ const getSharedMethods = (envs: TestResult["envs"]) => {
       getEnv(key, updatedEnvs),
       E.fromOption(() => "INVALID_KEY" as const),
 
-      E.map(({ value }) =>
+      E.map((env) =>
         pipe(
-          parseTemplateStringE(value, [
-            ...updatedEnvs.selected,
-            ...updatedEnvs.global,
-          ]),
-          // If the recursive resolution failed, return the unresolved value
-          E.getOrElse(() => value)
+          env,
+          E.fromNullable(undefined),
+          E.map((e) =>
+            pipe(
+              parseTemplateStringE(!e.secret ? e.value?.toString() : "", [
+                ...updatedEnvs.selected,
+                ...updatedEnvs.global,
+              ]),
+              // If the recursive resolution failed, return the unresolved value
+              E.getOrElse(() => (!e.secret ? e.value?.toString() : ""))
+            )
+          )
```

Comment:
```ts
      E.map(({ value }) =>
        pipe(
          parseTemplateStringE(value ?? "", [
            ...updatedEnvs.selected,
            ...updatedEnvs.global,
          ]),
          // If the recursive resolution failed, return the unresolved value
          E.getOrElse(() => value ?? "")
        )
      ),
```

Won't the `E.fromOption` above handle the case with invalid keys? Since secret environment variables are unwrapped before reaching the scripting context, we can specify value directly with a fallback to an empty string.